### PR TITLE
Add openshift to circleci

### PR DESCRIPTION
**Pull request by sampaccoud**
## Purpose

We want to be able to run our playbooks in the CI to really test deployments.

## Proposal

The easiest way to run OpenShift is to use the Docker image.
It requires a minor configuration of the Docker engine to allow an insecure registry.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -129,6 +129,27 @@ jobs:
...
+      - *attach_workspace
+      - *docker_load
+      - *ci_env
+      - run:
+          name: Test deploy in OpenShift
+          command: |
+            sudo bash -c "echo '{\"insecure-registries\": [\"172.30.0.0/16\"]}' > /etc/docker/daemon.json"
+            sudo service docker restart
+            wget https://github.com/openshift/origin/releases/download/v3.9.0/openshift-origin-client-tools-v3.9.0-191fece-linux-64bit.tar.gz
+            tar -xf openshift*.tar.gz --strip-components=1
```
Comment (jmaupetit) [inline .circleci/config.yml]: Why not using `tar xzf openshift*.tar.gz`? In a temporary directory and make an install (copy `oc`) in `$HOME/bin`?

File: .circleci/config.yml [hunk 2]
```diff
@@ -129,6 +129,27 @@ jobs:
...
+    steps:
+      - checkout
+      - *attach_workspace
+      - *docker_load
+      - *ci_env
+      - run:
+          name: Test deploy in OpenShift
+          command: |
+            sudo bash -c "echo '{\"insecure-registries\": [\"172.30.0.0/16\"]}' > /etc/docker/daemon.json"
+            sudo service docker restart
```
Comment (jmaupetit) [inline .circleci/config.yml]: I would prefer making the configuration after the installation of `oc`. WDYT?

File: .circleci/config.yml [hunk 3]
```diff
@@ -129,6 +129,27 @@ jobs:
...
+            wget https://github.com/openshift/origin/releases/download/v3.9.0/openshift-origin-client-tools-v3.9.0-191fece-linux-64bit.tar.gz
+            tar -xf openshift*.tar.gz --strip-components=1
+            export PATH="${PATH}:${PWD}"
+            LOCALE_IP=$(ip addr show eth0 | grep "inet\b" | awk '{print $2}' | cut -d/ -f1)
```
Comment (jmaupetit) [inline .circleci/config.yml]: Is the following easier to understand?

```bash
LOCALE_IP=$(ip -o addr show eth0 | grep -v inet6 | sed "s/^.* inet \([0-9.]*\)\/[0-9]* .*$/\1/")
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -129,6 +129,27 @@ jobs:
...
+            oc cluster up --server-loglevel=5 --public-hostname="${LOCALE_IP}"
+            oc login -u developer -p developer https://${LOCALE_IP}:8443 --insecure-skip-tls-verify=true
+            echo "arnold" | USE_TTY=false ./bin/bootstrap -e "apps_filter=richie"
```
Comment (jmaupetit) [inline .circleci/config.yml]: I think you can split this step in 3 separated steps:

1. Install & configure oc
2. Start the OpenShift cluster
3. Bootstrap the test application

File: bin/_config.sh [hunk 5]
```diff
@@ -60,25 +58,24 @@ function _docker_run() {
...
-    if ${OC_LOGIN} && ! minishift status | grep Running > /dev/null ; then
-        echo "Error: minishift is not running. Please start it first."
-        exit 1
-    fi
-
     # Check that the current user is already logged in minishift
-    if ${OC_LOGIN} && ! oc whoami &> /dev/null; then
-        echo "Error: you need to login to minishift first."
+    if ! oc whoami &> /dev/null; then
+        echo "Error: you need to login to openshift first."
```
Comment (jmaupetit) [inline bin/_config.sh]: OpenShift

File: bin/_config.sh [hunk 6]
```diff
@@ -60,25 +58,24 @@ function _docker_run() {
...
         exit 1
     fi
 
-    docker run --rm -it \
+    OPENSHIFT_URL=$(oc version|grep "Server "|awk -F'Server ' '{print $2}')
```
Comment (jmaupetit) [inline bin/_config.sh]: Can you add spaces between pipes to increase readability?

Comment (jmaupetit) [reply bin/_config.sh]: Default awk separator is space, hence, this row can be simplified as follow:

```bash
OPENSHIFT_URL=$(oc version | grep Server | awk '{print $2}')
```

File: bin/_config.sh [hunk 7]
```diff
@@ -60,25 +58,24 @@ function _docker_run() {
...
+
+    docker run --rm -i${TTY_OPTION} \
         -u "$(id -u)" \
         --env-file "$env_file" \
         --env OC_LOGIN \
         --env K8S_AUTH_API_KEY="$(oc whoami -t)" \
-        --env K8S_AUTH_HOST="https://$(minishift ip):8443" \
-        --env MINISHIFT_IP="$(minishift ip)" \
+        --env K8S_AUTH_HOST="${OPENSHIFT_URL}" \
+        --env OPENSHIFT_DOMAIN="${OPENSHIFT_DOMAIN}" \
```
Comment (jmaupetit) [inline bin/_config.sh]: If it's the same variable name, you can pass it to docker run with: `--env OPENSHIFT_DOMAIN`

**Review by jmaupetit — COMMENTED** [review]
That's an awesome feature. Thanks for having taken the time to implement it. :pray:

File: bin/_config.sh [hunk 8]
```diff
@@ -60,25 +58,28 @@ function _docker_run() {
         i=$(( i+1 ))
     done
 
-    # Check that minishift is running
-    if ${OC_LOGIN} && ! minishift status | grep Running > /dev/null ; then
-        echo "Error: minishift is not running. Please start it first."
+    # Check that the current user is already logged in OpenShift
+    if ! oc whoami &> /dev/null; then
```
Comment (jmaupetit) [inline bin/_config.sh]: Some scripts using Arnold (such as `pytest`) do not require to be logged in to OpenShift, and, thus are not impacted by the `K8S_AUTH_HOST` or `OPENSHIFT_DOMAIN` values. Hence, I think we must keep the `OC_LOGIN` flag functional:

```bash
if ${OC_LOGIN} && ! oc whoami &> /dev/null; then
    # ...
```

File: bin/_config.sh [hunk 9]
```diff
@@ -60,25 +58,28 @@ function _docker_run() {
...
+        echo "Error: you need to login to an OpenShift server first."
         exit 1
     fi
 
-    # Check that the current user is already logged in minishift
-    if ${OC_LOGIN} && ! oc whoami &> /dev/null; then
-        echo "Error: you need to login to minishift first."
-        exit 1
-    fi
+    [[ "$USE_TTY" == "false" ]] && TTY_OPTION="" || TTY_OPTION="t"
```
Comment (jmaupetit) [inline bin/_config.sh]: I would prefer: `TTY_OPTION="-t"` and adding a space between `-i` and the `${TTY_OPTION}` in the `docker run` invocation.

**Review by jmaupetit — APPROVED** [review]
I only have two minor changes (the latest is optional). After that: ship it!

File: bin/_config.sh (hunk 9 above)
Comment (sampaccoud) [reply bin/_config.sh]: ok

**Pull request merged**